[LightGBM] How to set predict_disable_shape_check=true  when prediction with lightgbm on mmlspark?

I'm always frustrated when I met the fatal error as below:
Caused by: java.lang.Exception: Booster Predict call failed in LightGBM with error: The number of features in data (224) is not the same as it was in training data (225).

Sometimes I can avoid this case through feature frequency filter, but when some small amount feature approaching threshold, the error raise up again, so this param config is very necessary;

Could someone help to add this feature ?
It's better to add an api for pass these k-v config to lightgbm, that would be awesome!!!

--- a/lightgbm/src/main/python/synapse/ml/lightgbm/mixin.py
+++ b/lightgbm/src/main/python/synapse/ml/lightgbm/mixin.py
@@ -64,3 +64,14 @@ class LightGBMModelMixin:
             The number of features.
         """
         return self._java_obj.getBoosterNumFeatures()
+
+    def setPredictDisableShapeCheck(self, value=None):
+        """
+        Set shape check or not when predict.
+        """
+        if not value:
+            value = False
+        else:
+            value = True
+
+        self._java_obj.setPredictDisableShapeCheck(value)

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMClassifier.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMClassifier.scala
@@ -163,11 +163,11 @@ class LightGBMClassificationModel(override val uid: String)
   override def numClasses: Int = getActualNumClasses
 
   override def predictRaw(features: Vector): Vector = {
-    Vectors.dense(getModel.score(features, true, true))
+    Vectors.dense(getModel.score(features, true, true, getPredictDisableShapeCheck))
   }
 
   override def predictProbability(features: Vector): Vector = {
-    Vectors.dense(getModel.score(features, false, true))
+    Vectors.dense(getModel.score(features, false, true, getPredictDisableShapeCheck))
   }
 
   override def copy(extra: ParamMap): LightGBMClassificationModel = defaultCopy(extra)

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMRanker.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMRanker.scala
@@ -144,7 +144,7 @@ class LightGBMRankerModel(override val uid: String)
 
   override def predict(features: Vector): Double = {
     logPredict(
-      getModel.score(features, false, false)(0)
+      getModel.score(features, false, false, getPredictDisableShapeCheck)(0)
     )
   }
 

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMRegressor.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/LightGBMRegressor.scala
@@ -127,7 +127,7 @@ class LightGBMRegressionModel(override val uid: String)
 
   override def predict(features: Vector): Double = {
     logPredict(
-      getModel.score(features, false, false)(0)
+      getModel.score(features, false, false, getPredictDisableShapeCheck)(0)
     )
   }
 

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/booster/LightGBMBooster.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/booster/LightGBMBooster.scala
@@ -387,14 +387,14 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
     }
   }
 
-  def score(features: Vector, raw: Boolean, classification: Boolean): Array[Double] = {
+  def score(features: Vector, raw: Boolean, classification: Boolean, disableShapeCheck: Boolean): Array[Double] = {
     val kind =
       if (raw) boosterHandler.rawScoreConstant
       else boosterHandler.normalScoreConstant
     features match {
-      case dense: DenseVector => predictForMat(dense.toArray, kind,
+      case dense: DenseVector => predictForMat(dense.toArray, kind, disableShapeCheck,
         boosterHandler.scoredDataLengthLongPtr.get().ptr, boosterHandler.scoredDataOutPtr.get().ptr)
-      case sparse: SparseVector => predictForCSR(sparse, kind,
+      case sparse: SparseVector => predictForCSR(sparse, kind, disableShapeCheck,
         boosterHandler.scoredDataLengthLongPtr.get().ptr, boosterHandler.scoredDataOutPtr.get().ptr)
     }
     predScoreToArray(classification, boosterHandler.scoredDataOutPtr.get().ptr, kind)
@@ -403,9 +403,9 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
   def predictLeaf(features: Vector): Array[Double] = {
     val kind = boosterHandler.leafIndexPredictConstant
     features match {
-      case dense: DenseVector => predictForMat(dense.toArray, kind,
+      case dense: DenseVector => predictForMat(dense.toArray, kind, false,
         boosterHandler.leafIndexDataLengthLongPtr.get().ptr, boosterHandler.leafIndexDataOutPtr.get().ptr)
-      case sparse: SparseVector => predictForCSR(sparse, kind,
+      case sparse: SparseVector => predictForCSR(sparse, kind, false,
         boosterHandler.leafIndexDataLengthLongPtr.get().ptr, boosterHandler.leafIndexDataOutPtr.get().ptr)
     }
     predLeafToArray(boosterHandler.leafIndexDataOutPtr.get().ptr)
@@ -414,9 +414,9 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
   def featuresShap(features: Vector): Array[Double] = {
     val kind = boosterHandler.contribPredictConstant
     features match {
-      case dense: DenseVector => predictForMat(dense.toArray, kind,
+      case dense: DenseVector => predictForMat(dense.toArray, kind, false,
         boosterHandler.shapDataLengthLongPtr.get().ptr, boosterHandler.shapDataOutPtr.get().ptr)
-      case sparse: SparseVector => predictForCSR(sparse, kind,
+      case sparse: SparseVector => predictForCSR(sparse, kind, false,
         boosterHandler.shapDataLengthLongPtr.get().ptr, boosterHandler.shapDataOutPtr.get().ptr)
     }
     shapToArray(boosterHandler.shapDataOutPtr.get().ptr)
@@ -508,11 +508,12 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
   lazy val numTotalIterations: Int = numTotalModel / numModelPerIteration
 
   protected def predictForCSR(sparseVector: SparseVector, kind: Int,
+                              disableShapeCheck: Boolean,
                               dataLengthLongPtr: SWIGTYPE_p_long_long,
                               dataOutPtr: SWIGTYPE_p_double): Unit = {
     val numCols = sparseVector.size
 
-    val datasetParams = "max_bin=255"
+    val datasetParams = s"max_bin=255 predict_disable_shape_check=${disableShapeCheck.toString}"
     val dataInt32bitType = boosterHandler.dataInt32bitType
     val data64bitType = boosterHandler.data64bitType
 
@@ -526,6 +527,7 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
   }
 
   protected def predictForMat(row: Array[Double], kind: Int,
+                              disableShapeCheck: Boolean,
                               dataLengthLongPtr: SWIGTYPE_p_long_long,
                               dataOutPtr: SWIGTYPE_p_double): Unit = {
     val data64bitType = boosterHandler.data64bitType
@@ -533,7 +535,7 @@ class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val para
     val numCols = row.length
     val isRowMajor = 1
 
-    val datasetParams = "max_bin=255"
+    val datasetParams = s"max_bin=255 predict_disable_shape_check=${disableShapeCheck.toString}"
 
     LightGBMUtils.validate(
       lightgbmlib.LGBM_BoosterPredictForMatSingle(

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/params/LightGBMParams.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/params/LightGBMParams.scala
@@ -258,7 +258,7 @@ trait LightGBMPredictionParams extends Wrappable {
   def getFeaturesShapCol: String = $(featuresShapCol)
   def setFeaturesShapCol(value: String): this.type = set(featuresShapCol, value)
 
-  val predictDisableShapeCheck = new Param[Boolean](this, "predictDisableShapeCheck",
+  val predictDisableShapeCheck = new BooleanParam(this, "predictDisableShapeCheck",
     "control whether or not LightGBM raises an error " +
       "when you try to predict on data with a different number of features than the training data")
   setDefault(predictDisableShapeCheck -> false)

--- a/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/params/LightGBMParams.scala
+++ b/lightgbm/src/main/scala/com/microsoft/azure/synapse/ml/lightgbm/params/LightGBMParams.scala
@@ -257,6 +257,14 @@ trait LightGBMPredictionParams extends Wrappable {
 
   def getFeaturesShapCol: String = $(featuresShapCol)
   def setFeaturesShapCol(value: String): this.type = set(featuresShapCol, value)
+
+  val predictDisableShapeCheck = new Param[Boolean](this, "predictDisableShapeCheck",
+    "control whether or not LightGBM raises an error " +
+      "when you try to predict on data with a different number of features than the training data")
+  setDefault(predictDisableShapeCheck -> false)
+
+  def getPredictDisableShapeCheck: Boolean = $(predictDisableShapeCheck)
+  def setPredictDisableShapeCheck(value: Boolean): this.type = set(predictDisableShapeCheck, value)
 }
 
 /** Defines parameters for LightGBM models

--- a/lightgbm/src/test/scala/com/microsoft/azure/synapse/ml/lightgbm/split1/VerifyLightGBMClassifier.scala
+++ b/lightgbm/src/test/scala/com/microsoft/azure/synapse/ml/lightgbm/split1/VerifyLightGBMClassifier.scala
@@ -14,8 +14,8 @@ import com.microsoft.azure.synapse.ml.stages.MultiColumnAdapter
 import org.apache.commons.io.FileUtils
 import org.apache.spark.TaskContext
 import org.apache.spark.ml.evaluation.{BinaryClassificationEvaluator, MulticlassClassificationEvaluator}
-import org.apache.spark.ml.feature.{StringIndexer, VectorAssembler}
-import org.apache.spark.ml.linalg.{DenseVector, Vector}
+import org.apache.spark.ml.feature.{LabeledPoint, StringIndexer, VectorAssembler}
+import org.apache.spark.ml.linalg.{DenseVector, Vector, Vectors}
 import org.apache.spark.ml.tuning.{ParamGridBuilder, TrainValidationSplit}
 import org.apache.spark.ml.util.MLReadable
 import org.apache.spark.ml.{Estimator, Model}
@@ -639,6 +639,31 @@ class VerifyLightGBMClassifier extends Benchmarks with EstimatorFuzzing[LightGBM
 
     // Validate fit works and doesn't get stuck
     assertFitWithoutErrors(baseModel, df)
+  }
+
+  test("Verify LightGBM Classifier won't get stuck on " +
+    "number of features in data is not the same as it was in training data") {
+    val inputData = Seq(
+      LabeledPoint(1.0, Vectors.dense(1.0, 0.0, 3.0)),
+      LabeledPoint(0.0, Vectors.sparse(3, Array(0, 2), Array(1.0, 3.0)))
+    )
+    val testData = Seq(
+      (
+        "uuid-value",
+        Vectors.sparse(15, Array(2, 6, 8, 14), Array(1.0, 1.0, 1.0, 2.0))
+      )
+    )
+
+    val modelDF = spark.createDataFrame(inputData).toDF("labels", "features")
+    val testDF = spark.createDataFrame(testData).toDF("uuid", "features")
+
+    val fitModel = baseModel.fit(modelDF)
+    val oldModelString = fitModel.getModel.modelStr.get
+
+    val testModel = LightGBMClassificationModel.loadNativeModelFromString(oldModelString)
+    testModel.setPredictDisableShapeCheck(true)
+
+    assert(testModel.transform(testDF).collect().length > 0)
   }
 
   def verifyLearnerOnBinaryCsvFile(fileName: String,

